sdk/metric/integrator/simple/simple.go

```diff
@@ -68,97 +84,271 @@ type (
-		AggregationSelector: selector,
-		stateful:            stateful,
-		batch: batch{
-			values:        map[batchKey]batchValue{},
-			intervalStart: time.Now(),
+		AggregationSelector: aselector,
+		ExportKindSelector:  eselector,
+		state: state{
+			values:        map[stateKey]*stateValue{},
+			processStart:  now,
+			intervalStart: now,
 		},
 	}
 }
 
-func (b *Integrator) Process(accumulation export.Accumulation) error {
+// Process implements export.Integrator.
+func (b *Integrator) Process(accum export.Accumulation) error {
 	if b.startedCollection != b.finishedCollection+1 {
 		return ErrInconsistentState
 	}
-
-	desc := accumulation.Descriptor()
-	key := batchKey{
+	desc := accum.Descriptor()
+	key := stateKey{
 		descriptor: desc,
-		distinct:   accumulation.Labels().Equivalent(),
-		resource:   accumulation.Resource().Equivalent(),
-	}
-	agg := accumulation.Aggregator()
-	value, ok := b.batch.values[key]
-	if ok {
-		// Note: The call to Merge here combines only
-		// identical accumulations.  It is required even for a
-		// stateless Integrator because such identical accumulations
-		// may arise in the Meter implementation due to race
-		// conditions.
-		return value.aggregator.Merge(agg, desc)
-	}
-	// If this integrator is stateful, create a copy of the
-	// Aggregator for long-term storage.  Otherwise the
-	// Meter implementation will checkpoint the aggregator
-	// again, overwriting the long-lived state.
-	if b.stateful {
-		tmp := agg
-		// Note: the call to AggregatorFor() followed by Merge
-		// is effectively a Clone() operation.
-		b.AggregatorFor(desc, &agg)
-		if err := agg.Merge(tmp, desc); err != nil {
-			return err
+		distinct:   accum.Labels().Equivalent(),
+		resource:   accum.Resource().Equivalent(),
+	}
+	agg := accum.Aggregator()
+
+	// Check if there is an existing value.
+	value, ok := b.state.values[key]
+	if !ok {
+		stateful := b.ExportKindFor(desc, agg.Aggregation().Kind()).MemoryRequired(desc.MetricKind())
+
+		newValue := &stateValue{
+			labels:   accum.Labels(),
+			resource: accum.Resource(),
+			updated:  b.state.finishedCollection,
+			stateful: stateful,
+			current:  agg,
 		}
+		if stateful {
+			if desc.MetricKind().PrecomputedSum() {
+				// If we know we need to compute deltas, allocate two aggregators.
+				b.AggregatorFor(desc, &newValue.cumulative, &newValue.delta)
+			} else {
+				// In this case we are not certain to need a delta, only allocate a
+				// cumulative aggregator.  We _may_ need a delta accumulator if
+				// multiple synchronous Accumulators produce an Accumulation (handled
+				// below), which requires merging them into a temporary Aggregator.
+				b.AggregatorFor(desc, &newValue.cumulative)
+			}
+		}
+		b.state.values[key] = newValue
+		return nil
+	}
+
+	// Advance the update sequence number.
+	sameCollection := b.state.finishedCollection == value.updated
+	value.updated = b.state.finishedCollection
+
+	// An existing value will be found for some stateKey when:
```

Comment:
You changed the terminology to be a bit clearer, but I still think this comment is wrong. We set `b.state.values[key] = newValue` on line 146 outside of the `if stateful { ... }` block, and indeed I believe we need to otherwise `value.current` would never be set (e.g. line 184).

Should this not just be "sameCollection will be true when: (b) ..."

Case (a) seems to be redundant for the code that follows (which to me seems to only be concerned with multi-accumulator aggregation).

Reply:
I can see how the comment is confusing. I meant "An existing value will be found _AT THIS POINT IN THE CODE_". To reach this point in the code, we have to have already passed through line 146 (in the current collection, if stateless, or in any previous collection, if not stateless). I added a point that (a) can be true if stateless, and (b) is true the second time through whether stateful or not. I'm struggling to improve on this...